ateam_common/src/km_assignment.cpp

```diff
@@ -0,0 +1,331 @@
+  const Eigen::MatrixXd & matrix)
+{
+  if (matrix.rows() == matrix.cols()) {
+    return matrix;
+  }
+
+  std::size_t new_dim = std::max(matrix.rows(), matrix.cols());
+
+  //  The value of any non-existant cost is 0, less than any assigable
+  //  real cost.
+  Eigen::MatrixXd new_mat =
+    Eigen::MatrixXd::Constant(new_dim, new_dim, 0);
+
+  new_mat.block(0, 0, matrix.rows(), matrix.cols()) = matrix;
+
+  return new_mat;
+}
+
+const Eigen::MatrixXd replace_nan_costs_with_zeros(
+  const Eigen::MatrixXd & matrix)
+{
+  Eigen::MatrixXd zero_mask = Eigen::MatrixXd::Zero(matrix.rows(), matrix.cols());
+  Eigen::MatrixXd new_matrix = (matrix.array().isNaN()).select(zero_mask, matrix);
+  return new_matrix;
+}
+
+const Eigen::MatrixXd replace_nan_costs_with_value(
+  const Eigen::MatrixXd & matrix,
+  double value)
+{
+  Eigen::MatrixXd value_mask = Eigen::MatrixXd::Ones(matrix.rows(), matrix.cols()) * value;
+  Eigen::MatrixXd new_matrix = (matrix.array().isNaN()).select(value_mask, matrix);
+  return new_matrix;
+}
+
+void compute_slack(
+  const int x,
+  std::vector<double> & slack,
+  std::vector<double> & slackx,
+  const Eigen::MatrixXd & cost,
+  const Eigen::VectorXd & lx,
+  const Eigen::VectorXd & ly
+)
+/*
+The slack used to update the labeling (weight) of nodes.
+
+- Initialized before starting the algorithm.
+- Called every time we add an augmenting path.
+- Called after each time we update labels using the minimum slack
+value.
+*/
+{
+  for (size_t y = 0; y < static_cast<size_t>(cost.cols()); ++y) {
+    if (lx[x] + ly[y] - cost(x, y) < slack[y]) {
+      slack[y] = lx[x] + ly[y] - cost(x, y);
+      slackx[y] = x;
+    }
+  }
+}
+
+std::vector<int> max_cost_assignment(
+  const Eigen::MatrixXd & cost_matrix,
+  bool max_cost
+)
+{
+  /*
+  Find the maximum cost assignment (aka min cost if you invert all the signs) of
+  a bipartite graph (i.e. robots and field positions) using the Hungarian (Kuhn-Munkres)
+  Maximum Matching Algorithm.
+
+  This implementation is largely based off of the one in dlib but allows
+  us to use our own types and types from Eigen rather than the dlib ones
+  as well as use fp (and not just int) values.
+  https://github.com/davisking/dlib/blob/master/dlib/optimization/max_cost_assignment.h
+
+  This resource can be used for reference:
+  (two different links, but the content is identical)
+      https://cse.hkust.edu.hk/~golin/COMP572/Notes/Matching.pdf
+      https://www.columbia.edu/~cs2035/courses/ieor6614.S16/GolinAssignmentNotes.pdf
+
+  This resource is a bit more clear but I unfortunately found it after the two above:
+    https://cp-algorithms.com/graph/hungarian-algorithm.html#implementation-of-the-hungarian-algorithm
+  */
+
+  // Make sure our matrix fits our algorithms requirements
+  // (it is square and non-negative).
+  auto cost = cost_matrix;
+  if (cost_matrix.cols() != cost_matrix.rows()) {
+    cost = make_square_cost_matrix(cost_matrix);
+  }
+  cost = scale_matrix(cost, max_cost);
```

Comment:
Do we need to flip the order of these steps?
If a non-square cost matrix comes in with both negative and positive numbers, the zeros added by `make_square_cost_matrix` will get shifted up to positive costs by `scale_matrix`. If the intent is that the zeros added by the squaring function are to indicate "don't pick me unless you have to," then we probably want to add those after the scaling shifts the costs to be non-negative.

Reply:
Yes, we do need to flip these! I addressed a similar issue with the order of implementation for dealing with NaN values but forgot to check it for the general case, and my test cases were simple enough that this didn't cause any issues.